### DISCUSSION TITLE
How to incorporate kopf as any other module in the python code

### DISCUSSION THREAD

**Discussion by jntupal** [nolar/kopf#1111] [Q&A]
Hi Community.

Brief about my foobar microservice:
I was trying to exercise the kopf in my python foobar microservice running in K8s. This service as such interacts with  network devices in south direction and in north side it's suppose to interface with CRD of kubernetes to get the input params.

Essentially, kopf will be my interface for interaction with CRD of K8s. I'm using kopf and running it as separate thread in my example snippet code file: foobar_k8s.py

def kopf_thread_loop():

    try:
        print("kopf_thread_loop started...")
        kopf.run()
    except Exception as e:
        print("Exception in kopf: %s", str(e))

In the Dockerfile of foobar microservice, I run the python program as:
CMD ["python3","main.py"]

main.py file and other files has the other important definitions for network device interaction (running as multi-threaded).

Although, with above I don't see any reception of create/update/delete etc.. of CR through kopf.

However, when I try using the below method in the dockerfile, I'm able to see the reception of create/update/delete CR.
CMD kopf run /foobar_k8s.py

Any comments or help on the above will be of great help.

Regards, Yogen

**Comment by nolar** [comment]
Hi. Do you import that main.py file in the script that runs `kopf.run()`? Those decorators need to be executed & the functions need to be declared to be runnable.

Also, do you see any starting logs when executed with `kopf.run()`?

**Reply by jntupal** [reply]
No, I don't import main.py in the script (foobar_k8.py). it's otherway for my case. So, main.py has the code for main() entry point and from there I call an initialize() method which internally calls the kopf_thread_loop(). See below my file flow:

main.py                          
===========
from foobar_k8 import  k8_init

main()
|--> initialize()
          |
          |-> k8_init()


foobar_k8s.py
=============
import kopf
import ...

def k8_init():

    # Start KOPF in a separate thread
    kopf_thread = threading.Thread(target = kopf_thread_loop,
                                   args = (foobar_ctx,))
    kopf_thread.daemon = True
    kopf_thread.start()
    return 0

def kopf_thread_loop(foobar_ctx):

try:
    print("kopf_thread_loop started...")
    kopf.run()
except Exception as e:
    print("Exception in kopf: %s", str(e))


NOTE: Also, I don't see any starting logs when I run kopf.run()

**Reply by nolar** [reply]
But where are all the handlers that should react to creation/update/deletion?

**Reply by jntupal** [reply]
All the handlers are residing in the same file foobar_k8.py
